File: algorytm.py
def start():
    pracownicy = [] #lista pracownikow
    zmiany = [] #lista zmiany
    zmianya = int(input("podaj ilosc zmian na dzien"))
    for i in range(1, zmianya * 8): #żeby symulacja było prostsza, bierzemy pod uwagę tydzień a nie miesiac
        zmiany.append(zmiana(i, [])) #wypelniam liste zmiany obiektami o klasie zmiana
    dodaj_pracownika(pracownicy, zmiany, zmianya)

class pracownik: #klasa pracownik a w niej imie, nazwisko, godziny tygodniowo i dziennie
    def __init__(self, imie, nazwisko, godziny_tydzien, godziny_zmiana, zmiana):
        self.imie = imie
        self.nazwisko = nazwisko
        self.godziny_tydzien = godziny_tydzien
        self.godziny_zmiana = godziny_zmiana
        self.zmiana = zmiana
    def __repr__(self): #tutaj definiuje co zwróci klasa kiedy ją się wywoła
        return f"{self.imie} {self.nazwisko} {self.godziny_tydzien} {self.godziny_zmiana} {self.zmiana}"

class zmiana: #klasa zmiana a w niej id i pracownicy ktorzy beda pracowac w tym dniu
    def __init__(self, id, pracownicy):
        self.id = id
        self.pracownicy = pracownicy
    def __repr__(self):
        return f"{self.id} {self.pracownicy}"
    
def alg(pracownicy, zmiany, ilosc_zmian):
    for pracownik in pracownicy:
        for x in zmiany:
            if pracownik.godziny_tydzien > 0:
                n = 0
                while True: 
                    dzielnik = pracownik.zmiana + ilosc_zmian * n
                    if x.id == dzielnik:
                        x.pracownicy.append([pracownik.imie, pracownik.godziny_zmiana]) #dodaje do listy pracownikow liste z imieniem pracownika i godzinami do przepracowania w dniu
                        #zmiany[x.id - 1] = zmiana(x.id, x.pracownicy) #do okreslonego przez numer iteracji elementu listy dodaje obiekt o klasie zmiana z tym samym id ale ze zmodyfikowaną listą
                        pracownik.godziny_tydzien -= pracownik.godziny_zmiana
                    n += 1
                    if dzielnik > x.id:
                        break

    result(zmiany)

def result(zmiany):
    print(zmiany) #printuje wynik

def dodaj_pracownika(pracownicy, zmiany, ilosc_zmian):
    while True:
        imie = input("imie:")
        if imie == "/q": #dla prostoty. mozna dodac zapytanie na koniec czy dodac nowego pracownika or smth
            break
        nazwisko = input("nazwisko:")
        godziny = int(input("godziny tygodniowo (pelen etat=40 g tyg):"))
        godziny_zmiana = int(input("godziny na zmiane:"))
        zmiana_ktora = int(input("na ktorą zmiane:"))
        pracownicy.append(pracownik(imie, nazwisko, godziny, godziny_zmiana, zmiana_ktora)) #do listy pracownicy dodaje obiekt o klasie pracownik o podanych parametrach
        #print(pracownicy)
        #pominąłem sprawdzanie czy pracownik juz istnieje ale mozna potem dodac
        #pominąłem mozliwosc wprowadzenia wielu zmian ale to mozna potem dodac
    alg(pracownicy, zmiany, ilosc_zmian)

start()

File: algorytm.cs
using System;
using System.Xml.XPath;

namespace algorytm
{
    internal class Program
    {

        public class Pracownik
        {
            public string Imie { get; set; }
            public string Nazwisko { get; set; }
            public int Godziny_tydzien { get; set; }
            public int Godziny_zmiana { get; set; }
            public int Zmiana { get; set; }

        public Pracownik(string imie, string nazwisko, int godziny_tydzien, int godziny_zmiana, int zmiana )
        {
            Imie = imie;
            Nazwisko = nazwisko;
            Godziny_tydzien = godziny_tydzien;
            Godziny_zmiana = godziny_zmiana;
            Zmiana = zmiana;
        }
        }   

        public class Zmiana
        {
            public int Id { get; set; }
            public List<Pracownik> Pracownicy { get; set; }

        public Zmiana(int id, List<Pracownik> pracownicy )
        {
            Id = id;
            Pracownicy = pracownicy;
        }

        }  
        static void Main()
        {
            List<Pracownik> pracownicy = new List<Pracownik>();
            List<Zmiana> zmiany = new List<Zmiana>();
            Console.Write("ile zmian na dzien?:");
            int zmianya = Int32.Parse(Console.ReadLine());
            for (int i = 1; i < zmianya * 8; i++)
            {
                zmiany.Add(new Zmiana(i, new List<Pracownik>()));
            }
            Dodaj_pracownika(pracownicy, zmiany, zmianya);
        }

        static void Dodaj_pracownika(List<Pracownik> pracownicy, List<Zmiana> zmiany, int ilosc_zmian)
        {
            while (true)
            {
                Console.Write("imie:");
                string imie = Console.ReadLine();
                if (imie == "/q")
                {
                    break;
                }
                Console.Write("nazwisko:");
                string nazwisko = Console.ReadLine();
                Console.Write("godziny tygodniowo(pelen etat=40g):");
                int godziny_tydzien = Int32.Parse(Console.ReadLine());
                Console.Write("godziny na zmiane:");
                int godziny_zmiana = Int32.Parse(Console.ReadLine());
                Console.Write("na ktora zmiane:");
                int zmiana = Int32.Parse(Console.ReadLine());
                pracownicy.Add(new Pracownik(imie, nazwisko, godziny_tydzien, godziny_zmiana, zmiana));
            }
            Alg(pracownicy, zmiany, ilosc_zmian);
        }

        static void Alg(List<Pracownik> pracownicy, List<Zmiana> zmiany, int ilosc_zmian)
        {
            foreach (Pracownik pracownik in pracownicy)
            {
                foreach (Zmiana zmiana in zmiany)
                {
                    if (pracownik.Godziny_tydzien > 0)
                    {
                        int n = 0;
                        while (true)
                        {
                            int dzielnik = pracownik.Zmiana + ilosc_zmian * n;
                            if (zmiana.Id == dzielnik)
                            {
                                zmiana.Pracownicy.Add(new Pracownik(pracownik.Imie, pracownik.Nazwisko, pracownik.Godziny_tydzien, pracownik.Godziny_zmiana, pracownik.Zmiana));
                                pracownik.Godziny_tydzien -= pracownik.Godziny_zmiana;
                            }
                            n += 1;
                            if (dzielnik > zmiana.Id)
                            {
                                break;
                            }
                        }
                    }
                }
            }
            Result(zmiany);
        }

        static void Result(List<Zmiana> zmiany)
        {
            foreach (Zmiana zmiana in zmiany)
        {
            Console.WriteLine($"id: {zmiana.Id}");
            Console.WriteLine("pracownicy na zmianie:");

            foreach (Pracownik pracownik in zmiana.Pracownicy)
            {
                Console.WriteLine($"  - {pracownik.Imie} {pracownik.Nazwisko} {pracownik.Godziny_zmiana}");
            }

            Console.WriteLine(); 
        }
        }
    }
}